radixdlt/src/main/java/com/radixdlt/consensus/VertexStore.java

```diff
@@ -17,168 +17,288 @@
-		if (tipVertex == null) {
-			throw new IllegalStateException("Committing a vertex which was never inserted: " + vertexId);
+	public void processLocalSync(Hash vertexId) {
+		syncing.remove(vertexId);
+	}
+
+	public boolean syncToQC(QuorumCertificate qc, QuorumCertificate committedQC, @Nullable ECPublicKey author) {
+		if (addQC(qc)) {
+			return true;
 		}
-		final LinkedList<Vertex> path = new LinkedList<>();
-		Vertex vertex = tipVertex;
-		while (vertex != null && !root.equals(vertex)) {
-			path.addFirst(vertex);
-			vertex = vertices.remove(vertex.getParentId());
+
+		if (author != null) {
+			this.doSync(qc, author);
 		}
 
-		for (Vertex committed : path) {
-			if (committed.getAtom() != null) {
-				CommittedAtom committedAtom = committed.getAtom().committed(commitMetadata);
-				storeAtom(committedAtom);
-			}
 
-			lastCommittedVertex.onNext(committed);
+		/*
+		Optional<VertexMetadata> committed = committedQC.getCommitted();
+		if (!committed.isPresent()) {
+			return false;
+		}
+
+		VertexMetadata committedMetadata = committed.get();
+		if (root.getView().compareTo(committedMetadata.getView()) < 0) {
+
+			long stateVersion = committed.get().getStateVersion();
+			final Hash hash;
+			try {
+				// TODO: Make it easier to retrieve signatures of QC
+				hash = Hash.of(DefaultSerialization.getInstance().toDson(committedQC.getVoteData(), Output.HASH));
+			} catch (SerializationException e) {
+				throw new IllegalStateException("Failed to serialize");
+			}				List<ECPublicKey> signers = committedQC.getSignatures().signedMessage(hash);
+
+			Observable.combineLatest(
+				vertexSupplier.sendGetVerticesRequest(committedQC.getProposed().getId(), author, 3).toObservable(),
+				syncedStateComputer.syncTo(stateVersion, signers).toSingleDefault(0).toObservable(),
+				(v, l) -> v
+			).subscribe(vertices -> {
+				synchronized (lock) {
+					if (root.getView().compareTo(committedMetadata.getView()) < 0) {
+						uncommitted.clear();
+						for (Vertex v : vertices) {
+							if (!v.getId().equals(committedMetadata.getId())) {
+								uncommitted.put(v.getId(), v);
+							}
+						}
+						this.root = committedMetadata;
+						this.highestCommittedQC = committedQC;
+						this.highestQC = committedQC;
+					}
+				}
+			});
 		}
+		*/
+
+		return false;
+	}
 
-		vertices.remove(root.getId());
-		root = tipVertex;
+	private boolean addQC(QuorumCertificate qc) {
+		synchronized (lock) {
+			if (!vertices.containsKey(qc.getProposed().getId())) {
+				return false;
+			}
+
+			if (highestQC.get().getView().compareTo(qc.getView()) < 0) {
+				highestQC.set(qc);
+			}
 
-		updateVertexStoreSize();
-		return tipVertex;
+			QuorumCertificate highestCommitted = highestCommittedQC.get();
+			qc.getCommitted().ifPresent(vertexMetadata -> {
+				if (!highestCommitted.getCommitted().isPresent()
+					|| highestCommitted.getCommitted().get().getStateVersion() < vertexMetadata.getStateVersion()) {
+					this.highestCommittedQC.set(qc);
+				}
+			});
```

Comment:
For an atomic version, this could be turned into something like:
```suggestion
			QuorumCertificate highestCommitted = highestCommittedQC.get();
			highestCommittedQC.getAndUpdate(currentHighest -> qc.getCommitted().map(vertexMetadata -> {
			if (!highestCommitted.getCommitted().isPresent()
				|| highestCommitted.getCommitted().get().getStateVersion() < vertexMetadata.getStateVersion()) {
				return qc;
			} else {
				return currentHighest;
			}
		}).orElse(currentHighest));
```

Reply:
Urgh, the formatting of the suggested change here is a bit awful. My suggestion is essentially just moving the get + update wholly inside AtomicReference's own getAndUpdate.

Reply:
this no longer needs to be atomically updated since only one thread is ever accessing VertexStore at a time.